smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/commit/KafkaThrottledLatestProcessedCommit.java

```diff
@@ -0,0 +1,143 @@
+package io.smallrye.reactive.messaging.kafka.commit;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import io.smallrye.reactive.messaging.kafka.IncomingKafkaRecord;
+import io.vertx.kafka.client.common.TopicPartition;
+import io.vertx.kafka.client.consumer.OffsetAndMetadata;
+import io.vertx.mutiny.core.Context;
+import io.vertx.mutiny.kafka.client.consumer.KafkaConsumer;
+
+public class KafkaThrottledLatestProcessedCommit implements KafkaCommitHandler {
+
+    private static final long THROTTLE_TIME_IN_MILLIS = 5_000L;
+    private static final Map<String, Map<Integer, TopicPartition>> TOPIC_PARTITIONS_CACHE = new ConcurrentHashMap<>();
+
+    private final Map<TopicPartition, Queue<Long>> receivedOffsetQueues = new HashMap<>();
+    private final Map<TopicPartition, Set<Long>> processedOffsetSets = new HashMap<>();
+
+    private final KafkaConsumer<?, ?> consumer;
+
+    private volatile Context context;
+    private long nextCommitTime;
+
+    public KafkaThrottledLatestProcessedCommit(KafkaConsumer<?, ?> consumer) {
+        this.consumer = consumer;
+    }
+
+    private <K, V> TopicPartition getTopicPartition(IncomingKafkaRecord<K, V> record) {
+        return TOPIC_PARTITIONS_CACHE
+                .computeIfAbsent(record.getTopic(), topic -> new ConcurrentHashMap<>())
+                .computeIfAbsent(record.getPartition(), partition -> new TopicPartition(record.getTopic(), partition));
+    }
+
+    private Set<Long> getProcessedOffsetSet(TopicPartition topicPartition) {
+        return processedOffsetSets
+                .computeIfAbsent(topicPartition, t -> new HashSet<>());
+    }
+
+    private Queue<Long> getReceivedOffsetQueue(TopicPartition topicPartition) {
+        return receivedOffsetQueues
+                .computeIfAbsent(topicPartition, t -> new LinkedList<>());
+    }
+
+    @Override
+    public void partitionsAssigned(Context context, Set<TopicPartition> partitions) {
+        this.context = context;
+
+        receivedOffsetQueues.clear();
+        processedOffsetSets.clear();
+
+        resetNextCommitTime();
+    }
+
+    @Override
+    public <K, V> IncomingKafkaRecord<K, V> received(IncomingKafkaRecord<K, V> record) {
+        TopicPartition recordsTopicPartition = getTopicPartition(record);
+        getReceivedOffsetQueue(recordsTopicPartition).offer(record.getOffset());
+
+        return record;
+    }
```

Comment:
This solves the issue with compacted logs. Tested and working well, but it creates a new problem. 

The user might decide to use the **NONE** Acknowledgement Strategy together with **enable.auto.commit=false**.
https://smallrye.io/smallrye-reactive-messaging/2.1.1/apidocs/org/eclipse/microprofile/reactive/messaging/Acknowledgment.Strategy.html#NONE

 Basically the user wishes to never commit (there's a few use cases for this).

But with the implementation as is we will spring a leak. The received offset queue will grow without being emptied since there will be no call to **ack**. It looks to me that the Incoming publishers and connectors are decoupled in such a way that the connector doesn't know what strategy is configured.

I see two possible solutions:

1. Have the commit strategy be configurable like the nack strategy. Have the default depend on **enable.auto.commit** but let the user configure _ignore_ if they wish.

2. If we have more than **max.poll.records** (default 500) received messages without committing a single one then we can assume that the user configured to never commit. In this case we can just turn the strategy off.

1 is much simpler to maintain, but is a bit onerous on the user and requires proper documentation.
2 requires no action from the user (should just work), but is a harder to reason and maintain.


What do you think? If there was some way for the connector to know that **NONE** was used during setup that would be great, but I'm not seeing it.

Reply:
(1) definitely. We can use "enable.auto.commit" to determine the default. Adding another attribute to configure the strategy would be nice and allow configuring the application to use the current model. 

That's what I wrote on the SmallRye Mailing list (for traceability purpose):

> To implement this, we would need a new attribute to set the strategy:
> * commit every offset when the message is acked -> Today's solution
> * let kafka auto-commit -> enable the auto-commit and ignore acks from the app
> * batch/cursor tracking strategy -> the strategy described in this mail (which would need a > better name)
> Each of them has pros and cons, so they would need to be clearly documented.